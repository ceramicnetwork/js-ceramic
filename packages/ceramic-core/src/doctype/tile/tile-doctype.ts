import CID from 'cids'

import jsonpatch from 'fast-json-patch'

import { encode as base64Encode } from '@ethersproject/base64'
import { randomBytes } from '@ethersproject/random'

import { DocState, Doctype, InitOpts } from "../../doctype"
import User from "../../user"
import { Context } from "../../context"

const DOCTYPE = 'tile'

/**
 * Tile doctype parameters
 */
export interface TileParams {
    content: object;
    owners?: Array<string>;
}

/**
 * Tile doctype implementation
 */
export class TileDoctype implements Doctype {
    id: string;
    content: object
    doctype: string
    head: CID
    owners: Array<string>
    state: DocState

    /**
     * Create Tile doctype
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: TileParams, context: Context, opts?: InitOpts): Promise<TileDoctype> {
        if (context.user == null) {
            throw new Error('No user authenticated')
        }

        const { content, owners } = params
        const record = await TileDoctype.makeGenesis(context.user, content, owners, opts)

        const tileDoctype = await context.ceramic.createFromGenesis(record, opts)
        return Promise.resolve(tileDoctype)
    }

    /**
     * Make genesis record
     * @param user - User instance
     * @param content - Context
     * @param owners - Owners
     * @param opts - Initialization parameters
     */
    static async makeGenesis(user: User, content: any, owners?: Array<string>, opts: { isUnique?: boolean } = {}): Promise<any> {
        if (!user) {
            throw new Error('No user authenticated')
        }

        if (!owners) {
            owners = [user.DID]
        }

        let unique: string
        if (opts.isUnique) {
            unique = base64Encode(randomBytes(12))
        }

        const record = { doctype: DOCTYPE, owners, content, unique }
        return TileDoctype._signRecord(record, user)
    }

    /**
     * Change existing Tile doctype
     * @param doctype - Tile doctype instance
     * @param params - Change parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async change(doctype: Doctype, params: TileParams, context: Context, opts?: InitOpts): Promise<Doctype> {
        if (context.user == null) {
            throw new Error('No user authenticated')
        }

        const updateRecord = TileDoctype._makeRecord(doctype, context.user, params.content)
        return await context.ceramic.applyRecord(doctype.id, updateRecord, opts)
    }

    /**
     * Make change record
     * @param doctype - Tile doctype instance
     * @param user - User instance
     * @param newContent - New context
     * @private
     */
    static async _makeRecord(doctype: Doctype, user: User, newContent: any): Promise<Doctype> {
        if (user == null) {
            throw new Error('No user authenticated')
        }

        const patch = jsonpatch.compare(doctype.content, newContent)
        const record: any = { owners: doctype.owners, content: patch, prev: doctype.head, id: doctype.state.log[0] }
        record.iss = user.DID
        // convert CID to string for signing
        const tmpPrev = record.prev
        const tmpId = record.id
        record.prev = { '/': tmpPrev.toString() }
        record.id = { '/': tmpId.toString() }
        const jwt = await user.sign(record, { useMgmt: true})
        const [header, payload, signature] = jwt.split('.') // eslint-disable-line @typescript-eslint/no-unused-vars
        record.prev = tmpPrev
        record.id = tmpId

        return { ...record, header, signature }
    }

    /**
     * Sign Tile record
     * @param user - User instance
     * @param record - Record to be signed
     * @private
     */
    static async _signRecord(record: any, user: User): Promise<any> {
        if (user == null) {
            throw new Error('No user authenticated')
        }
        // TODO - use the dag-jose library for properly encoded signed records
        // The way we use did-jwts right now are quite hacky
        record.iss = user.DID
        // convert CID to string for signing
        const tmpCID = record.prev
        const tmpId = record.id
        if (tmpCID) {
            record.prev = { '/': tmpCID.toString() }
        }
        if (tmpId) {
            record.id = { '/': tmpId.toString() }
        }
        const jwt = await user.sign(record)
        const [header, payload, signature] = jwt.split('.') // eslint-disable-line @typescript-eslint/no-unused-vars
        if (tmpCID) {
            record.prev = tmpCID
        }
        if (tmpId) {
            record.id = tmpId
        }
        return { ...record, header, signature }
    }

}

