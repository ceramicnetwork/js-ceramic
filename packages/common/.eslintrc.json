{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint", "import"],
  "rules": {
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/ban-ts-ignore": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "import/no-unresolved": "error",
    "import/default": "off"
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "typescript": {
        "extensions": []
      },
      "node": true
    }
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/typescript"
  ]
}
