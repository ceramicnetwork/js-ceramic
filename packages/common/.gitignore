node_modules/
coverage/
lib/
src/generated/
