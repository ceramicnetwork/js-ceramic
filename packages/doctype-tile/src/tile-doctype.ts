import jsonpatch from 'fast-json-patch'

import { encode as base64Encode } from '@ethersproject/base64'
import { randomBytes } from '@ethersproject/random'

import { DID } from 'dids'
import {
    Doctype,
    DoctypeConstructor,
    DoctypeStatic,
    DocOpts,
    DocParams,
    Context
} from "@ceramicnetwork/common"

const DOCTYPE = 'tile'

/**
 * Tile doctype parameters
 */
export interface TileParams extends DocParams {
    content?: object;
}

/**
 * Tile doctype implementation
 */
@DoctypeStatic<DoctypeConstructor<TileDoctype>>()
export class TileDoctype extends Doctype {

    /**
     * Change existing Tile doctype
     * @param params - Change parameters
     * @param opts - Initialization options
     */
    async change(params: TileParams, opts: DocOpts = {}): Promise<void> {
        if (this.context.did == null) {
            throw new Error('No DID authenticated')
        }

        const updateRecord = await TileDoctype._makeRecord(this, this.context.did, params.content, params.metadata?.controllers, params.metadata?.schema)
        const updated = await this.context.api.applyRecord(this.id.toString(), updateRecord, opts)
        this.state = updated.state
    }

    /**
     * Create Tile doctype
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    static async create(params: TileParams, context: Context, opts?: DocOpts): Promise<TileDoctype> {
        if (context.did == null) {
            throw new Error('No DID authenticated')
        }

        const { content, metadata } = params
        const record = await TileDoctype.makeGenesis({ content, metadata }, context)
        return context.api.createDocumentFromGenesis<TileDoctype>(DOCTYPE, record, opts)
    }

    /**
     * Creates genesis record
     * @param params - Create parameters
     * @param context - Ceramic context
     */
    static async makeGenesis(params: DocParams, context?: Context): Promise<Record<string, any>> {
        const metadata = params.metadata? params.metadata : { controllers: [] }

        // check for DID and authentication
        if (!context.did || !context.did.authenticated) {
            throw new Error('No DID authenticated')
        }

        let unique: string
        if (params.deterministic) {
            unique = '0'
        } else {
            // If 'deterministic' is undefined, default to creating document uniquely
            unique = base64Encode(randomBytes(12))
        }

        const { controllers } = metadata
        if (!controllers || controllers.length === 0) {
            metadata.controllers = [context.did.id]
        }

        const { content } = params
        const record = { data: content, header: metadata, unique }
        return content ? TileDoctype._signDagJWS(record, context.did, metadata.controllers[0]) : record
    }

    /**
     * Make change record
     * @param doctype - Tile doctype instance
     * @param did - DID instance
     * @param newContent - New context
     * @param newControllers - New controllers
     * @param schema - New schema ID
     * @private
     */
    static async _makeRecord(doctype: Doctype, did: DID, newContent: any, newControllers?: string[], schema?: string): Promise<any> {
        if (did == null || !did.authenticated) {
            throw new Error('No DID authenticated')
        }

        const header: Record<string, any> = {}
        if (schema) {
            header.schema = schema
        }

        if (newControllers) {
            header.controllers = newControllers
        }

        const nonce = TileDoctype._calculateNonce(doctype)
        if (nonce != null) {
            header.nonce = nonce
        }

        if (newContent == null) {
            newContent = doctype.content
        }

        const patch = jsonpatch.compare(doctype.content, newContent)

        const willSquash = header.nonce && header.nonce > 0
        const prev = doctype.state.log[doctype.state.log.length - 1 - (willSquash ? 1 : 0)].cid

        const record = { header, data: patch, prev, id: doctype.state.log[0].cid }
        return TileDoctype._signDagJWS(record, did, doctype.controllers[0])
    }

    /**
     * Calculates anchor nonce
     */
    private static _calculateNonce(doctype: Doctype): number {
        // if there hasn't been any update prior to this we should set the nonce to 0
        if (!doctype.state.next) {
            return null
        }
        // get the current nonce and increment it by one
        return (doctype.state.next.metadata?.nonce || 0) + 1
    }

    /**
     * Sign Tile record
     * @param did - DID instance
     * @param record - Record to be signed
     * @param controller - Controller
     * @private
     */
    static async _signDagJWS(record: any, did: DID, controller: string): Promise<any> {
        if (did == null || !did.authenticated) {
            throw new Error('No user authenticated')
        }
        return did.createDagJWS(record, { did: controller })
    }

}
