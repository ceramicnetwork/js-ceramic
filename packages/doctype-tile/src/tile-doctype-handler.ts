import CID from 'cids'

import * as didJwt from 'did-jwt'
import base64url from 'base64url'
import { PublicKey } from 'did-resolver'

import jsonpatch from 'fast-json-patch'
import cloneDeep from 'lodash.clonedeep'

import { TileDoctype, TileParams } from "./tile-doctype"
import {
    AnchorProof,
    AnchorRecord,
    AnchorStatus,
    Context,
    DocOpts,
    DocState,
    RecordType,
    DoctypeConstructor,
    DoctypeHandler,
    DoctypeUtils,
    SignatureStatus
} from "@ceramicnetwork/common"

const DOCTYPE = 'tile'

/**
 * Tile doctype handler implementation
 */
export class TileDoctypeHandler implements DoctypeHandler<TileDoctype> {
    /**
     * Gets doctype name
     */
    get name(): string {
        return DOCTYPE
    }

    /**
     * Gets doctype class
     */
    get doctype(): DoctypeConstructor<TileDoctype> {
        return TileDoctype
    }

    /**
     * Create new Tile doctype instance
     * @param params - Create parameters
     * @param context - Ceramic context
     * @param opts - Initialization options
     */
    async create(params: TileParams, context: Context, opts?: DocOpts): Promise<TileDoctype> {
        return TileDoctype.create(params, context, opts)
    }

    /**
     * Applies record (genesis|signed|anchor)
     * @param record - Record
     * @param cid - Record CID
     * @param context - Ceramic context
     * @param state - Document state
     */
    async applyRecord(record: any, cid: CID, context: Context, state?: DocState): Promise<DocState> {
        if (state == null) {
            // apply genesis
            return this._applyGenesis(record, cid, context)
        }

        if (record.proof) {
            const proofRecord = (await context.ipfs.dag.get(record.proof)).value;
            return this._applyAnchor(record, proofRecord, cid, state);
        }

        return this._applySigned(record, cid, state, context)
    }

    /**
     * Applies genesis record
     * @param record - Genesis record
     * @param cid - Genesis record CID
     * @param context - Ceramic context
     * @private
     */
    async _applyGenesis(record: any, cid: CID, context: Context): Promise<DocState> {
        let payload = record
        const isSigned = DoctypeUtils.isSignedRecord(record)
        if (isSigned) {
            payload = (await context.ipfs.dag.get(record.link)).value
            await this._verifySignature(record, context, payload.header.controllers[0])
        } else if (payload.data !== null) {
            throw Error('Genesis record with contents should always be signed')
        }
        return {
            doctype: DOCTYPE,
            content: payload.data,
            metadata: payload.header,
            signature: isSigned? SignatureStatus.SIGNED : SignatureStatus.GENESIS,
            anchorStatus: AnchorStatus.NOT_REQUESTED,
            log: [{ cid, type: RecordType.GENESIS }]
        }
    }

    /**
     * Applies signed record
     * @param record - Signed record
     * @param cid - Signed record CID
     * @param state - Document state
     * @param context - Ceramic context
     * @private
     */
    async _applySigned(record: any, cid: CID, state: DocState, context: Context): Promise<DocState> {
        await this._verifySignature(record, context, state.metadata.controllers[0])

        const payload = (await context.ipfs.dag.get(record.link)).value
        if (!payload.id.equals(state.log[0].cid)) {
            throw new Error(`Invalid docId ${payload.id}, expected ${state.log[0].cid}`)
        }

        const nextState = cloneDeep(state)

        nextState.signature = SignatureStatus.SIGNED
        nextState.anchorStatus = AnchorStatus.NOT_REQUESTED

        const nonce = payload.header.nonce
        const squash = nonce > 0 && state.next
        if (squash) {
            nextState.log[nextState.log.length-1] = { cid, type: RecordType.SIGNED }
            nextState.next = {
                content: jsonpatch.applyPatch(state.next.content, payload.data).newDocument
            }
        } else {
            nextState.log.push({ cid, type: RecordType.SIGNED })
            nextState.next = {
                content: jsonpatch.applyPatch(state.content, payload.data).newDocument
            }
        }

        if (payload.header) {
            nextState.next.metadata = {...nextState.metadata, ...payload.header}
        }

        if (nonce) {
            nextState.next.metadata = { ...nextState.next.metadata, nonce: nonce }
        }
        return nextState
    }

    /**
     * Applies anchor record
     * @param record - Anchor record
     * @param proof - Anchor record proof
     * @param cid - Anchor record CID
     * @param state - Document state
     * @private
     */
    async _applyAnchor(record: AnchorRecord, proof: AnchorProof, cid: CID, state: DocState): Promise<DocState> {
        state.log.push({ cid, type: RecordType.ANCHOR })
        let content = state.content
        let metadata = state.metadata

        if (state.next?.content) {
            content = state.next.content
            delete state.next.content
        }

        if (state.next?.metadata) {
            metadata = state.next.metadata
            delete state.next.metadata
        }

        delete state.next

        return {
            ...state, content, metadata, anchorStatus: AnchorStatus.ANCHORED, anchorProof: proof,
        }
    }

    /**
     * Verifies record signature
     * @param record - Record to be verified
     * @param context - Ceramic context
     * @param did - DID value
     * @private
     */
    async _verifySignature(record: any, context: Context, did: string): Promise<void> {
        const { payload, signatures } = record
        const { signature,  protected: _protected } = signatures[0]

        const decodedHeader = JSON.parse(base64url.decode(_protected))
        const { kid } = decodedHeader
        if (!kid.startsWith(did)) {
            throw new Error(`Signature was made with wrong DID. Expected: ${did}, got: ${kid}`)
        }

        const { publicKey } = await context.resolver.resolve(kid)
        const jws = [_protected, payload, signature].join('.')
        try {
            await this.verifyJWS(jws, publicKey)
        } catch (e) {
            throw new Error('Invalid signature for signed record. ' + e)
        }
    }

    /**
     * Verifies JWS token
     * @param jws - JWS token
     * @param pubkeys - public key(s)
     */
    async verifyJWS(jws: string, pubkeys: PublicKey[]): Promise<void> {
        await didJwt.verifyJWS(jws, pubkeys)
    }

}
