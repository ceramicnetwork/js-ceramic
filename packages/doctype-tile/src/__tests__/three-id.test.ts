import CID from 'cids'

import { Resolver } from "did-resolver"

import dagCBOR from "ipld-dag-cbor"

import KeyDidResolver from '@ceramicnetwork/key-did-resolver'
import ThreeIdResolver from '@ceramicnetwork/3id-did-resolver'

import { DID } from 'dids'

import { AnchorRecord, Context, SignedRecordDTO } from "@ceramicnetwork/common"
import { TileDoctypeHandler } from "../tile-doctype-handler"
import { TileDoctype } from "../tile-doctype"
jest.mock('did-jwt', () => ({
  // TODO - We should test for when this function throws as well
  verifyJWS: (): void => { return }
}))

const cloneDeep = require('lodash.clonedeep') // eslint-disable-line @typescript-eslint/no-var-requires
const { sha256 } = require('js-sha256') // eslint-disable-line @typescript-eslint/no-var-requires
const hash = (data: string): CID => new CID(1, 'sha2-256', Buffer.from('1220' + sha256(data), 'hex'))

const FAKE_CID_1 = new CID('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = new CID('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = new CID('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = new CID('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')

const RECORDS = {
  genesis: { header: { tags: ['3id'], controllers: [ 'did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV' ] }, data: { publicKeys: { test: '0xabc' } } },
  genesisGenerated: {
    jws: {
      payload: "bbbb",
      signatures: [
        {
          protected: "eyJraWQiOiJkaWQ6a2V5OnpRM3Nod3NDZ0ZhbkJheDZVaWFMdTFvR3ZNN3ZodXFvVzg4VkJVaVVUQ2VIYlRlVFYiLCJhbGciOiJFUzI1NksifQ",
          signature: "cccc"
        }
      ],
      link: new CID("bafyreia4yhrqevqys35mvccnfmgcrpne6fxrdpqaonif7njq3gh7so633e")
    },
    linkedBlock: {
      data: {
          publicKeys: {
            test: "0xabc",
          },
      },
      header: {
        controllers: [
          "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV"
        ],
        "tags": [
          "3id",
        ]
      },
      unique: "0",
    }
  },
  r1: {
    desiredContent: { publicKeys: { test: '0xabc' }, other: 'data' },
    record: {
      jws: {
        "payload": "bbbb",
        "signatures": [
          {
            "protected": "eyJraWQiOiJkaWQ6a2V5OnpRM3Nod3NDZ0ZhbkJheDZVaWFMdTFvR3ZNN3ZodXFvVzg4VkJVaVVUQ2VIYlRlVFYiLCJhbGciOiJFUzI1NksifQ",
            "signature": "cccc"
          }
        ],
        "link": "bafyreib2rxk3rybk3aobmv5cjuql3bm2twh4jo5uxgf5kpqcsgz7soitae"
      },
      payload: {
        "id": "bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu",
        "data": [
          {
            "op": "add",
            "path": "/other",
            "value": "data"
          }
        ],
        "prev": "bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu",
        "header": {},
      }
    }
  },
  r2: { record: { proof: FAKE_CID_4 } },
  proof: {
    blockNumber: 123456,
    chainId: 'fakechain:123',
  }
}

const serialize = (data: any): any => {
  if (Array.isArray(data)) {
    const serialized = []
    for (const item of data) {
      serialized.push(serialize(item))
    }
    return serialized
  }
  if (!CID.isCID(data) && typeof data === "object") {
    const serialized: Record<string, any> = {}
    for (const prop in data) {
      serialized[prop] = serialize(data[prop])
    }
    return serialized
  }
  if (CID.isCID(data)) {
    return data.toString()
  }
  return data
}

describe('ThreeIdHandler', () => {
  let did: DID
  let tileDoctypeHandler: TileDoctypeHandler
  let context: Context

  beforeAll(() => {
    did = new DID()
    did.createJWS = jest.fn(async () => {
      // fake jws
      return { payload: 'bbbb', signatures: [{ protected: 'eyJraWQiOiJkaWQ6a2V5OnpRM3Nod3NDZ0ZhbkJheDZVaWFMdTFvR3ZNN3ZodXFvVzg4VkJVaVVUQ2VIYlRlVFYiLCJhbGciOiJFUzI1NksifQ', signature: 'cccc'}]}
    })
    did._id = 'did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV'

    const recs: Record<string, any> = {}
    const ipfs = {
      dag: {
        put(rec: any, cid?: CID): any {
          if (cid) {
            recs[cid.toString()] = { value: rec }
            return cid
          }
          // stringify as a way of doing deep copy
          const clone = cloneDeep(rec)
          const c = hash(JSON.stringify(clone))
          recs[c.toString()] = { value: clone }
          return c
        },
        get(cid: any): any {
          return recs[cid.toString()]
        }
      }
    }

    const threeIdResolver = ThreeIdResolver.getResolver({
      loadDocument: (): any => {
        return Promise.resolve({
          content: {
            "publicKeys": {
              "signing": "zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
              "encryption": "z6LSfQabSbJzX8WAm1qdQcHCHTzVv8a2u6F7kmzdodfvUCo9"
            }
          }
        })
      },
      createDocument: (): any => { return null },
    })

    const api = { getSupportedChains: jest.fn(async () => {return ["fakechain:123"]}) }
    const keyDidResolver = KeyDidResolver.getResolver()
    context = {
      did,
      ipfs,
      resolver: new Resolver({
        ...threeIdResolver, ...keyDidResolver
      }),
      anchorService: null,
      api,
    }
  })

  beforeEach(() => {
    tileDoctypeHandler = new TileDoctypeHandler()
  })

  it('is constructed correctly', async () => {
    expect(tileDoctypeHandler.name).toEqual('tile')
  })

  it('makes genesis record correctly', async () => {
    const record = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: RECORDS.genesis.header, deterministic:true }, context) as SignedRecordDTO
    const { jws, linkedBlock } = record

    const payload = dagCBOR.util.deserialize(linkedBlock)
    const generated = { jws: serialize(jws), linkedBlock: serialize(payload)}
    expect(generated).toEqual(serialize(RECORDS.genesisGenerated))
  })

  it('applies genesis record correctly', async () => {
    const tileHandler = new TileDoctypeHandler()

    const record = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id], tags: ['3id'] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(record, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(record.linkedBlock)
    await context.ipfs.dag.put(payload, record.jws.link)

    const docState = await tileHandler.applyRecord(record.jws, FAKE_CID_1, context)
    expect(docState).toMatchSnapshot()
  })

  it('makes signed record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    await context.ipfs.dag.put(RECORDS.genesisGenerated.jws, FAKE_CID_1)
    await context.ipfs.dag.put(RECORDS.genesisGenerated.linkedBlock, RECORDS.genesisGenerated.jws.link)

    const state = await tileDoctypeHandler.applyRecord(RECORDS.genesisGenerated.jws, FAKE_CID_1, context)
    const doctype = new TileDoctype(state, context)

    await expect(TileDoctype._makeRecord(doctype, null, RECORDS.r1.desiredContent)).rejects.toThrow(/No DID/)

    const record = await TileDoctype._makeRecord(doctype, did, RECORDS.r1.desiredContent) as SignedRecordDTO
    const { jws: rJws, linkedBlock: rLinkedBlock} = record
    const rPayload = dagCBOR.util.deserialize(rLinkedBlock)
    expect({ jws: serialize(rJws), payload: serialize(rPayload)}).toEqual(RECORDS.r1.record)
  })

  it('applies signed record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id], tags: ['3id'] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)

    // apply genesis
    let state = await tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)

    const doctype = new TileDoctype(state, context)
    const signedRecord = await TileDoctype._makeRecord(doctype, did, RECORDS.r1.desiredContent) as SignedRecordDTO

    await context.ipfs.dag.put(signedRecord, FAKE_CID_2)

    const sPayload = dagCBOR.util.deserialize(signedRecord.linkedBlock)
    await context.ipfs.dag.put(sPayload, signedRecord.jws.link)

    // apply signed
    state = await tileDoctypeHandler.applyRecord(signedRecord.jws, FAKE_CID_2, context, state)
    expect(state).toMatchSnapshot()
  })

  it('throws error if record signed by wrong DID', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: ['did:3:fake'], tags: ['3id'] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)

    await expect(tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)).rejects.toThrow(/wrong DID/)
  })

  it('applies anchor record correctly', async () => {
    const tileDoctypeHandler = new TileDoctypeHandler()

    const genesisRecord = await TileDoctype.makeGenesis({ content: RECORDS.genesis.data, metadata: { controllers: [did.id], tags: ['3id'] } }, context) as SignedRecordDTO
    await context.ipfs.dag.put(genesisRecord, FAKE_CID_1)

    const payload = dagCBOR.util.deserialize(genesisRecord.linkedBlock)
    await context.ipfs.dag.put(payload, genesisRecord.jws.link)

    // apply genesis
    let state = await tileDoctypeHandler.applyRecord(genesisRecord.jws, FAKE_CID_1, context)

    const doctype = new TileDoctype(state, context)
    const signedRecord = await TileDoctype._makeRecord(doctype, did, RECORDS.r1.desiredContent) as SignedRecordDTO

    await context.ipfs.dag.put(signedRecord, FAKE_CID_2)

    const sPayload = dagCBOR.util.deserialize(signedRecord.linkedBlock)
    await context.ipfs.dag.put(sPayload, signedRecord.jws.link)

    // apply signed
    state = await tileDoctypeHandler.applyRecord(signedRecord.jws, FAKE_CID_2, context, state)

    await context.ipfs.dag.put(RECORDS.proof, FAKE_CID_4)
    // apply anchor
    state = await tileDoctypeHandler.applyRecord(RECORDS.r2.record as AnchorRecord, FAKE_CID_3, context, state)
    expect(state).toMatchSnapshot()
  })
})
