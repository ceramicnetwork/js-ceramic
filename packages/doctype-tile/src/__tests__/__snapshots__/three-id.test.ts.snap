// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeIdHandler applies anchor record correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
  },
  "anchorStatus": 3,
  "content": Object {
    "other": "data",
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          119,
          169,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 2,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
    ],
    "tags": Array [
      "3id",
    ],
  },
  "signature": 2,
}
`;

exports[`ThreeIdHandler applies genesis record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
    ],
    "tags": Array [
      "3id",
    ],
  },
  "signature": 2,
}
`;

exports[`ThreeIdHandler applies signed record correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "publicKeys": Object {
      "test": "0xabc",
    },
  },
  "doctype": "tile",
  "log": Array [
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
    Object {
      "cid": Object {
        "codec": "dag-pb",
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          115,
          137,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 1,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
    ],
    "tags": Array [
      "3id",
    ],
  },
  "next": Object {
    "content": Object {
      "other": "data",
      "publicKeys": Object {
        "test": "0xabc",
      },
    },
    "metadata": Object {
      "controllers": Array [
        "did:key:zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV",
      ],
      "tags": Array [
        "3id",
      ],
    },
  },
  "signature": 2,
}
`;
